File: 2025-algorithm-design/lab/01/assets/main.py
def allocate_books(pages, m):
    def check(lim):
        num, tot = 0, 0
        for page in pages:
            if tot + page > lim:
                num += 1
                tot = 0
            tot += page
        if tot > 0:
            num += 1
        return num <= m

    lb, rb = max(pages), sum(pages)
    while lb < rb:
        mid = (lb + rb) // 2
        if check(mid):
            rb = mid
        else:
            lb = mid + 1
    return lb


if __name__ == "__main__":
    pages = [12, 34, 67, 90]
    m = 2
    print(allocate_books(pages, m))  # Output: 113

File: 2025-algorithm-design/lab/01/assets/main.lua
function AllocateBooks(pages, m)
	local function check(limit)
		local num, sum = 0, 0
		for i = 1, #pages do
			if sum + pages[i] > limit then
				num = num + 1
				sum = 0
			end
			sum = sum + pages[i]
		end
		if sum > 0 then
			num = num + 1
		end
		return num <= m
	end

	local l, r = 0, 2^31 - 1
	while l < r do
		local mid = math.floor((l + r) / 2)
		if check(mid) then
			r = mid
		else
			l = mid + 1
		end
	end

	print("The answer is " .. l)
end

local pages = {12, 34, 67, 90}
local m = 2
AllocateBooks(pages, m)
